From 9ee12278e230a99389a7b56530505d72f2f8ee7c Mon Sep 17 00:00:00 2001
From: Helmut Buchsbaum <helmut.buchsbaum@opensource.tttech-industrial.com>
Date: Wed, 17 Jul 2019 19:42:44 +0200
Subject: [PATCH 03/18] dm: acpi: validate minimal length of PSDS

Kontron COMe-mAL10 does not provide an ACPI PSDS table as expected.
basl_fwrite_psds() expects to copy 33 bytes beginning at offset 46, thus
accessing 79 bytes. The PSDS table provided here is shorter and thus
should be ignored. So introduce a minimum length check to achieve this.

Signed-off-by: Helmut Buchsbaum <helmut.buchsbaum@opensource.tttech-industrial.com>
---
 devicemodel/hw/platform/acpi/acpi.c | 17 ++++++++++++++---
 1 file changed, 14 insertions(+), 3 deletions(-)

diff --git a/devicemodel/hw/platform/acpi/acpi.c b/devicemodel/hw/platform/acpi/acpi.c
index 3d815975..82811655 100644
--- a/devicemodel/hw/platform/acpi/acpi.c
+++ b/devicemodel/hw/platform/acpi/acpi.c
@@ -1200,6 +1200,7 @@ acpi_build(struct vmctx *ctx, int ncpu)
 {
 	int err;
 	int i;
+	struct stat psds_stat;
 
 	basl_ncpu = ncpu;
 
@@ -1221,9 +1222,19 @@ acpi_build(struct vmctx *ctx, int ncpu)
 	err = basl_make_templates();
 
 	/* Check PSDS in SOS, only present PSDS table to UOS when PSDS present in SOS */
-	psds_fd = open("/sys/firmware/acpi/tables/PSDS", O_RDONLY);
-	if (psds_fd >=0)
-		acpi_table_enable(PSDS_ENTRY_NO);
+	err = stat("/sys/firmware/acpi/tables/PSDS", &psds_stat);
+	if (err == 0 && psds_stat.st_size >= 79) {
+		psds_fd = open("/sys/firmware/acpi/tables/PSDS", O_RDONLY);
+		if (psds_fd >=0)
+			acpi_table_enable(PSDS_ENTRY_NO);
+	} else {
+		if (err)
+			fprintf(stderr, "ACPI PSDS table not available, skipping\n");
+		else
+			fprintf(stderr, "ACPI PSDS table too small (%lu < 79), skipping\n",
+				psds_stat.st_size);
+		err = 0;
+	}
 
 	/*
 	 * Run through all the ASL files, compiling them and
-- 
2.20.1

